### PR TITLE
Modify HTTP frontend to return error code reflecting Triton error. Add test

### PR DESCRIPTION
**Pull request by GuanLuo**

### CODE REVIEW

File: src/sagemaker_server.cc [hunk 1]
```diff
@@ -407,49 +407,27 @@ SagemakerAPIServer::SagemakeInferRequestClass::InferResponseComplete(
...
 #endif  // TRITON_ENABLE_TRACING
 
-  if (err == nullptr) {
-    evthr_defer(infer_request->thread_, OKReplyCallback, infer_request);
-  } else {
+  if (err != nullptr) {
     EVBufferAddErrorJson(infer_request->req_->buffer_out, err);
+    // [FIXME] In http_server.cc, error handling is enhanced to reporting
+    // different error code according to the Triton error code, holding
+    // the change from SageMaker endpoint as it may not fit their SLA.
```
Comment (Tabrizian) [inline src/sagemaker_server.cc:414]: do we need similar fix for vertex endpoint as well?

**Review by Tabrizian — DISMISSED** [review]

File: qa/L0_http/http_test.py [hunk 2]
```diff
@@ -205,6 +207,31 @@ def test_content_encoding_unsupported_client(self):
...
+        # aligns with error detail (server unavailable).
+        t = threading.Thread(
+            target=self._raw_binary_helper, args=(model, input_bytes, input_bytes)
+        )
+        t.start()
+        time.sleep(0.5)
+        try:
+            self._raw_binary_helper(model, input_bytes, input_bytes)
+        except requests.exceptions.HTTPError as ex:
+            self.assertEqual(
```
Comment (rmccorm4) [inline qa/L0_http/http_test.py:225]: 1. Should the test be changed to expect exception to occur everytime? Looks like current test would pass if no exception occurs.
2. If so, will 0.5secs be flaky?

File: src/sagemaker_server.cc (hunk 1 above)
Comment (GuanLuo) [reply src/sagemaker_server.cc:414]: I will leave the same comment in vertex endpoint. Will also need to check with the right group of people for the same before applying the change

File: qa/L0_http/http_test.py (hunk 2 above)
Comment (GuanLuo) [reply qa/L0_http/http_test.py:225]: > Should the test be changed to expect exception to occur everytime? Looks like current test would pass if no exception occurs.

Good catch, will fix to fail if there is no exception

File: src/sagemaker_server.cc (hunk 1 above)
Comment (GuanLuo) [reply src/sagemaker_server.cc:414]: Actually it is harder for vertex AI.. The endpoint just call into the http_server base class's functions and it gets the same behavior already.. I would just let it be what it is and fix later if specific error code must be returned in vertex AI deployment.

**Review by rmccorm4 — APPROVED** [review]
LGTM assuming vertex/sagemaker OK

**Pull request merged**